15 座席表示・確保機能

fu10ta opened the pull request.

Commit:
add:座席表示の作成

用意した座席レイアウトを呼び出し映画タイトルと開始時刻・終了時刻とともに表示

Commit:
座席表示機能の実装

予約済みの席の反映
選択した席の色付け
再度選択で選択状態解除

Commit:
ボタン色の変更

Commit:
座席予約機能の実装

Commit:
Merge branch 'develop' of https://github.com/labotinc/codegym-team5 into feature/hinappu-15/#89

kazuyanakata reviewed (commented):
今のままだと、なんの機能をレビューしているのかパッと見で分からないので、タイトルの修正を他のを参考にお願いします。

kazuyanakata commented on html/mycakeapp/src/Template/Reserves/seat.ctp:
```diff
@@ -0,0 +1,24 @@
+<?php
+$startdate = $seatDetails['start_date'];
+$finishdate = date('G:i', strtotime('+' . $seatDetails['movie']['screening_time'] . 'minute', strtotime($startdate)));
+?>
```
変数の代入等はビューではなくコントローラーでやるべきなので、
reservesコントローラーで処理を行い、変数を渡す形に変更お願いします。

fu10ta replied:
対応いたしました。

kazuyanakata commented on html/mycakeapp/webroot/css/seat.css:
```diff
@@ -0,0 +1,72 @@
+}
+.seat_select li {
+    display: grid;
+    grid-template-columns: 1fr 1fr 1fr 1fr 1fr 1fr 1fr 1fr 1fr 1fr 1fr 1fr 1fr;
```
修正不要
```suggestion
    grid-template-columns: repeat(13, 1fr);
```
こちらの方が簡潔にかけます。

fu10ta replied:
対応いたしました。

kazuyanakata commented on html/mycakeapp/src/Controller/ReservesController.php:
```diff
@@ -33,32 +34,68 @@ public function initialize()
+        }
+        $this->viewBuilder()->setLayout('frame-title');
+        $title = '座席指定';
+        $this->set(compact('title', 'seatDetails', 'Reserved'));
```
座席予約数が0の時に 
compact(): Undefined variable: Reserved [APP/Controller/ReservesController.php, line 94]
とエラーが出るので対応お願いします。
おそらく、$seatReservationsが0の時の対応を書いて$Reservedに何かを入れれば大丈夫かと

fu10ta replied:
対応いたしました。

kazuyanakata commented on html/mycakeapp/src/Controller/ReservesController.php:
```diff
@@ -33,32 +34,68 @@ public function initialize()
+        $seatDetails = $this->Schedules->get($_SESSION['schedule_id'], [
+            'contain' => ['Movies', 'Theaters']
+        ]);
+        //２時間前までの席予約情報を取得
+        $seatReservations = $this->SeatReservations->find()
+            ->where(['schedule_id' => $_SESSION['schedule_id']])
+            ->andWhere(['is_cancelled' => 0])
+            ->andWhere(['created_at >=' => new Time('2 hours ago')]);
```
基本設計書には何も書いてなかったのですが、2時間前としているのは、どのような理由でしょうか？
以前に2時間について何か話していたのならすみません。
ただ、2時間前までに限定する必要はないかと！

fu10ta replied:
対応いたしました。

kazuyanakata commented on html/mycakeapp/src/Controller/ReservesController.php:
```diff
@@ -33,32 +34,68 @@ public function initialize()
             }
             $this->request->session()->delete('seat');
         }
+        if ($this->request->isPost()) {
```
①開発者ツールを用いて、予期せぬ値に変更した時の対応をお願いします。
　例1 シアターに無い座席番号や数字を入力する
　　2 全く違う値に変更する
　　3 既に予約済みの座席を指定する
②上記①が発生した時はエラーが発生している旨のエラー文を載せると良いと思います。

fu10ta replied:
③の状態のみ対応し、①②の場合に対して調べたいことが多いので一旦提出させていただきます。

kazuyanakata commented on html/mycakeapp/src/Template/Reserves/seat.ctp:
```diff
@@ -0,0 +1,24 @@
+<div class="seat_select">
+    <?= $this->element('Theater/' . $seatDetails['theater']['file_name']); ?>
+</div><!-- /.seat_select -->
+<?= $this->Form->button('次へ', [
+    'type' => 'submit',
+    'class' => 'normal-button back-orange',
+    'disabled' => 'true',
+]); ?>
```
1.グッドポイント
座席が選択されると次へボタンが黄色になる表示方法とても良いと思いました！
2.修正ポイント
何も選択されずにグレー状態でも、カーソルがポインターになり、影が表示されるのでそこは消した方がいいと思います！

fu10ta replied:
対応いたしました。

kazuyanakata commented on html/mycakeapp/webroot/css/seat.css:
```diff
@@ -0,0 +1,72 @@
+    height: 24px;
+    position: absolute;
+}
+.seat_select .reserved::before {
+    background-color: black;
+    content: "×";
+    color: white;
+    line-height: 24px;
+}
```
1.グッドポイント
既に予約されているものの表示方法わかりやすいです！
2.修正ポイント
選択不可のはずなのに、カーソルを載せるとポインターになり背景色に変わってしまうので、ここは変化なしが良いを思います！

fu10ta replied:
対応いたしました。

kazuyanakata commented on html/mycakeapp/src/Template/Reserves/seat.ctp:
```diff
@@ -0,0 +1,24 @@
+</script>
+<div class="seat_heading">
+    <h1><?= $seatDetails['movie']['name']; ?></h1>
+    <h2><?= $startdate->format('m月d日 G:i') ?>〜<span><?= $finishdate ?></span></h2>
```
曜日が抜けてます！
対応お願いします！

fu10ta replied:
対応いたしました。

kazuyanakata reviewed (commented):
レビュー完了しました！
修正お願いいたします！！

また、AppControllerの76行目のコメントを外すこともお願いします。

terukyou commented on html/mycakeapp/src/Template/Element/Theater/theater1.ctp:
```diff
@@ -0,0 +1,36 @@
+<?php
+//　補足：管理者画面作成時には値を変更することで対応可能
+$record_name = [1, 2, 3, 4, 5, 6, 7, 8];
+$column_name = ['A', 'B', 'C', 'D', 'E', 'F', 'G', 'H', 'I', 'J', 'K'];
+?>
+
+<ul>
+    <li>
+        <p></p> <!-- gridでの左上の空白 -->
+        <?php
+        //カラムヘッダー出力
+        for ($i = 0; $i < count($column_name); $i++) {
+            echo '<p>' . $column_name[$i] . '</p>';
+        }
```
> // 大文字のアルファベット A-Z を順番に出力
for ( $i = 0; $i < 26; $i++ ) { // アルファベットの数(=26)だけ処理ループ
	print chr(65 + $i) . '<br />';
}

引用元)http://tsunagu-tsunagaru.hatenablog.jp/entry/PHP-アルファベットA-Z/a-zを順番に出力する
for文にそのまま出力した方がシンプルですし、必要なアルファベットを全てコードに書く必要がなくなるかと思います。

fu10ta replied:
管理者画面作成を考えた場合に任意の文字列でシアターレイアウトを作成できるよう対応いたしましたため、このような形になっておりますが、一つのシアターレイアウトに対応する形で簡素化するか別途相談させていただきたいです。

terukyou replied:
修正不要にします!!(リファクタリングの修正依頼で緊急ではないため)

terukyou commented on html/mycakeapp/src/Template/Element/Theater/theater1.ctp:
```diff
@@ -0,0 +1,36 @@
+    </li>
+    <?php
+    // レコードヘッダーと席のボタン出力
+    for ($j = 0; $j < count($record_name); $j++) {
+        echo '<li>';
+        echo '<p>' . $record_name[$j] . '</p>';
+        for ($k = 0; $k < count($column_name); $k++) {
```
こちらも同様、iかjの値をそのまま出力か、変換して出力すると良いかと!!

fu10ta replied:
上記の理由と同様の種子で作成しております。

terukyou replied:
こちらも上記と同様に修正不要です！

terukyou commented on html/mycakeapp/src/Template/Reserves/seat.ctp:
```diff
@@ -0,0 +1,24 @@
+<?= $this->HTML->css('seat') ?>
+<?php
+$startdate = $seatDetails['start_date'];
+$finishdate = date('G:i', strtotime('+' . $seatDetails['movie']['screening_time'] . 'minute', strtotime($startdate)));
```
[Movie.php](https://github.com/labotinc/codegym-team5/commit/69332b64a895329dd7b5857255061de4662ca138)に共通関数として残しているので、ReservesController.phpの198付近を参考に利用していただければと思います。

fu10ta replied:
取得方法がいまいちわからず、一旦ベタ書きにて対応いたしました。

terukyou replied:
申し訳ないです🙇‍♂️🙇‍♂️
共通関数の書く場所を間違えており、こちらを使用すると間違った値が出てくるのでベタ書きしていただいて大丈夫です。

terukyou commented on html/mycakeapp/webroot/js/seat.js:
```diff
@@ -0,0 +1,25 @@
+    //予約済みのbutonにdisabled属性を追加
+    if(typeof Reserved != 'undefined'){
+        for(var i = 0; i < Object.keys(Reserved).length; i++){
+            console.log(Reserved);
```
こちらは不要かと思います。

fu10ta replied:
削除いたしました。　

Commit:
レビュー修正

・変数宣言をコントローラーへ
・CSS指定の簡潔化
・エラー対応
・予約情報の取得条件変更
・予約済みの席の対応
・予約状況や選択状態のCSS指定変更
・曜日表示追加
・不要な指定削除

ghost commented:
いただきましたレビューの修正をいたしました。
現状リクエスト時の対応が完璧でないことと、テスト仕様書を鑑みると穴が多い気がします。
一旦中田さんよりレビューいただいている処理は調べ対応いたしますので、一旦提出させていただきます。

Commit:
Auth処理追加

ghost commented:
> レビュー完了しました！
> 修正お願いいたします！！
> 
> また、AppControllerの76行目のコメントを外すこともお願いします。

kazuyanakata commented on html/mycakeapp/src/Controller/ReservesController.php:
```diff
@@ -54,8 +54,28 @@ public function seat()
+            $Reserved[] = $seatReservation['column_number'] . '-' . $seatReservation['record_number'];
+        }
+        //予約済み情報を取得
+        $Reservations = $this->ReservationDetails->find()
+            ->where(['schedule_id' => $_SESSION['schedule_id']])
+            ->andWhere(['is_cancelled' => 0]);
+        foreach ($Reservations as $Reservation) {
+            $Reserved[] = $Reservation['column_number'] . '-' . $Reservation['record_number'];
+        }
```
※緊急修正ではありません。
席予約情報のDBに無くて予約詳細情報のDBに有るデータはないはずので
ここは不要かと思います。

fu10ta replied:
そうですね！席予約に関して時間制限かけていたので、二度とりにいってましたが、一つで良いですね。

kazuyanakata commented on html/mycakeapp/src/Controller/ReservesController.php:
```diff
@@ -33,32 +34,76 @@ public function initialize()
+        }
+        if ($this->request->isPost()) {
+            $seatNumber = explode("-", $_POST['seat']);
+            if (in_array($_POST['seat'], $Reserved)) {
+                return $this->redirect(['controller' => 'error']);
+            }
```
※緊急修正ではありません
エラーに飛ばすなら、セッションは削除した方が良いと思います。
```suggestion
            if (in_array($_POST['seat'], $Reserved)) {
            　   $this->request->session()->delete('schedule_id');
                return $this->redirect(['controller' => 'error']);
            }
```

kazuyanakata commented on html/mycakeapp/src/Controller/ReservesController.php:
```diff
@@ -33,32 +34,76 @@ public function initialize()
+        }
+        if ($this->request->isPost()) {
+            $seatNumber = explode("-", $_POST['seat']);
+            if (in_array($_POST['seat'], $Reserved)) {
```
※緊急の修正ではないです
以下の場合にエラーが発生します。
①AさんがA-1の座席を予約した
②AさんがA-1の座席をキャンセルした
③AさんがA-1の再度座席を予約した

原因は、post送信した時にレコードの新規登録をする処理のみになっているからだと思います。
なので
①主キーを検索して該当するレコードがあるか
②あるならキャンセルフラグを外す
③ないなら新規登録する

という処理にしたら良いと思います。

terukyou replied:
私も確認し、解決方法を模索しましたが、ticket.ctpの戻るボタンを変更し、リンク先を指定したaタグにしたところ問題が治りました。
推測ですが、戻るボタンのhistory.back()が前の画面に戻る機能なので「まっさらなseatページ」でなく「post送信の履歴が残ったseatページ」になった結果、挙動がおかしくなるのだと思います。

kazuyanakata replied:
今ticketの方をHtmlヘルパーのリンクに直したらpost送信の履歴消えました！
緊急ブランチ作ってその箇所の修正これから行います！！

ただ、一度登録して削除したものを再登録すると画面では問題ないように見えましたが、DBのキャンセルフラグは今のままだとはずれてないので84行目に
$entity['is_cancelled'] = 0;
の追加をお願いします。

kazuyanakata replied:
ticket.ctpの方変更完了しましたので
developブランチをもう一度こちらに持ってきて対応ください！

kazuyanakata reviewed (commented):
お疲れ様です！
レビュー完了いたしました！
UIもバッチリですしコードもとても読みやすかったです！！

3点緊急ではないですがコメント残しましたので確認お願いいたします！
既にマージしても機能としては大きく問題ないと思いますが、平山さんレビュー中とかに対応いただけたらと思います！

terukyou commented on html/mycakeapp/src/Controller/ReservesController.php:
```diff
@@ -33,32 +34,76 @@ public function initialize()
+        $title = '座席指定';
+        $week = ['日', '月', '火', '水', '木', '金', '土'];
+        $startdate = $seatDetails['start_date']->format('m月d日　G:i') . '(' . $week[$seatDetails['start_date']->format('w')] . ')';
+        $finishdate = date('G:i', strtotime('+' . $seatDetails['movie']['screening_time'] . 'minute', strtotime($startdate)));
```
終了時刻がずれているので修正お願いします🙇‍♂️🙇‍♂️
(チケット選択ページの終了時刻と比較すると分かりやすいと思います!)
```suggestion
        $finishdate = date('G:i', strtotime('+' . $seatDetails['movie']['screening_time'] . 'minute', strtotime($seatDetails['start_date'])));
```

kazuyanakata committed:
fix:コントローラー修正

・データセーブ時にis_cancelledフラグを追加
・終了時刻の修正
